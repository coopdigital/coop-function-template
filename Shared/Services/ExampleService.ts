import { Service } from 'typedi';
import ExampleRepository from '../Dal/ExampleRepository';
import { ExampleDto } from '../Dtos/ExampleDto';
import { ExampleModel } from '../Models/ExampleModel';

/** Example Service, abstract business logic from the Function */
@Service()
export default class ExampleService {
  /** Initialises the Service */
  constructor(private exampleRepository: ExampleRepository) {}

  /** Gets a example record by id */
  public async get(id: string): Promise<ExampleDto> {
    let response: ExampleDto = null;

    const exampleData = await this.exampleRepository.getExampleById(id);

    if (exampleData) {
      response = {
        Field1: exampleData.Field1,
        Field2: exampleData.Field2,
      };
    }

    return response;
  }

  /** Updates an example */
  public async update(id: string, example: ExampleDto): Promise<void> {
    const exampleModel: ExampleModel = { id, ...example };

    await this.exampleRepository.updateExampleModel(exampleModel);
  }
}
