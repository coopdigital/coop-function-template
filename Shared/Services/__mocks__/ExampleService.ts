export const mockGet = jest.fn((id) => {
  return {
    Field1: 'Test Data',
    Field2: 1234,
  };
});

export const mockUpdate = jest.fn();

const mock = jest.fn().mockImplementation(() => {
  return { get: mockGet, update: mockUpdate };
});
export default mock;
