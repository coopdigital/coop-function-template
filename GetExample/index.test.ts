import getExampleTrigger from './index';
import Context from '../testing/defaultContext';
// @ts-ignore
import { mockGet } from '../Shared/Services/ExampleService';
jest.mock('../Shared/Services/ExampleService.ts');
jest.mock('typedi');

describe('Get Example Tests', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Success', async () => {
    // Arrange
    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
    };

    // Act
    await getExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(200);
    expect(testContext.res.body).toMatchSnapshot();
  });

  it('400 - Invalid Id', async () => {
    // Arrange
    const testContext = new Context();
    const testData = {
      params: {
        id: 'TestId',
      },
    };

    // Act
    await getExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(400);
  });

  it('404 - Invalid Id', async () => {
    // Arrange
    (mockGet as jest.Mock).mockImplementation(() => {
      return null;
    });

    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
    };

    // Act
    await getExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(404);
  });

  it('500 - Exception', async () => {
    // Arrange
    (mockGet as jest.Mock).mockImplementation(() => {
      throw new Error('TEST ERROR');
    });

    const testContext = new Context();
    const testData = {
      params: {
        id: 'd8db7f54-7758-4a5a-bfd1-82e205c80a59',
      },
    };

    // Act
    await getExampleTrigger(testContext, testData);

    // Assert
    expect(testContext.res.status).toBe(500);
  });
});
