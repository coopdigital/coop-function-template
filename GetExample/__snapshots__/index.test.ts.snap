// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Example Tests Success 1`] = `
Object {
  "Field1": "Test Data",
  "Field2": 1234,
}
`;
